File: server/test/alda/parser/variables_test.clj
(ns alda.parser.variables-test
  (:require [clojure.test      :refer :all]
            [alda.parser-util  :refer (parse-to-lisp-with-context)]
            [instaparse.core   :refer (failure?)]))

(deftest variable-name-tests
  (testing "variable names"
    (testing "must start with two letters"
      (is (= '((alda.lisp/get-variable :aa))
             (parse-to-lisp-with-context :music-data "aa")))
      (is (= '((alda.lisp/get-variable :aaa))
             (parse-to-lisp-with-context :music-data "aaa")))
      (is (= '((alda.lisp/get-variable :HI))
             (parse-to-lisp-with-context :music-data "HI")))
      (is (thrown? Exception (parse-to-lisp-with-context :music-data "x")))
      (is (thrown? Exception (parse-to-lisp-with-context :music-data "y2")))
      (is (thrown? Exception (parse-to-lisp-with-context :music-data "1234kittens")))
      (is (thrown? Exception (parse-to-lisp-with-context :music-data "r2d2")))
      (is (thrown? Exception (parse-to-lisp-with-context :music-data "i_like_underscores"))))
    (testing "can't contain pluses or minuses"
      (is (thrown? Exception (parse-to-lisp-with-context :music-data "jar-jar-binks")))
      (is (thrown? Exception (parse-to-lisp-with-context :music-data "han+leia")))
      (is (thrown? Exception (parse-to-lisp-with-context :music-data "ionlyprograminc++"))))
    (testing "can contain digits"
      (is (= '((alda.lisp/get-variable :celloPart2))
             (parse-to-lisp-with-context :music-data "celloPart2")))
      (is (= '((alda.lisp/get-variable :xy42))
             (parse-to-lisp-with-context :music-data "xy42")))
      (is (= '((alda.lisp/get-variable :my20cats))
             (parse-to-lisp-with-context :music-data "my20cats"))))
    (testing "can contain underscores"
      (is (= '((alda.lisp/get-variable :apple_cider))
             (parse-to-lisp-with-context :music-data "apple_cider")))
      (is (= '((alda.lisp/get-variable :underscores__are___great____))
             (parse-to-lisp-with-context :music-data "underscores__are___great____"))))))

(deftest variable-get-tests
  (testing "variable getting"
    (is (= '(alda.lisp/score
              (alda.lisp/part {:names ["flute"]}
                (alda.lisp/note (alda.lisp/pitch :c))
                (alda.lisp/get-variable :flan)
                (alda.lisp/note (alda.lisp/pitch :f))))
           (parse-to-lisp-with-context :score "flute: c flan f")))
    (is (= '(alda.lisp/score
              (alda.lisp/part {:names ["clarinet"]}
                (alda.lisp/get-variable :pudding123)))
           (parse-to-lisp-with-context :score "clarinet: pudding123")))))

(deftest variable-set-tests
  (testing "variable setting"
    (testing "within an instrument part"
      (is (= '(alda.lisp/score
                (alda.lisp/part {:names ["harpsichord"]}
                  (alda.lisp/set-variable :custard_
                    (alda.lisp/note (alda.lisp/pitch :c))
                    (alda.lisp/note (alda.lisp/pitch :d))
                    (alda.lisp/chord
                      (alda.lisp/note (alda.lisp/pitch :e))
                      (alda.lisp/note (alda.lisp/pitch :g))))))
             (parse-to-lisp-with-context :score "harpsichord:\n\ncustard_ = c d e/g")))
      (is (= '(alda.lisp/score
                (alda.lisp/part {:names ["glockenspiel"]}
                  (alda.lisp/set-variable :sorbet
                    (alda.lisp/note (alda.lisp/pitch :c))
                    (alda.lisp/note (alda.lisp/pitch :d))
                    (alda.lisp/chord
                      (alda.lisp/note (alda.lisp/pitch :e))
                      (alda.lisp/note (alda.lisp/pitch :g))))
                  (alda.lisp/note (alda.lisp/pitch :c))))
             (parse-to-lisp-with-context :score "glockenspiel:\n\nsorbet=c d e/g\nc"))))
    (testing "at the top of a score"
      (is (= '(alda.lisp/score
                (alda.lisp/set-variable :GELATO
                  (alda.lisp/note (alda.lisp/pitch :d))
                  (alda.lisp/note (alda.lisp/pitch :e)))
                (alda.lisp/part {:names ["clavinet"]}
                  (alda.lisp/chord
                    (alda.lisp/note (alda.lisp/pitch :c))
                    (alda.lisp/note (alda.lisp/pitch :f)))))
             (parse-to-lisp-with-context :score "GELATO=d e\n\nclavinet: c/f"))))))

File: client/parser/variables_test.go
package parser

import (
	"fmt"
	"testing"

	"alda.io/client/model"
	_ "alda.io/client/testing"
)

func variableDefinition(name string, events ...model.ScoreUpdate,
) model.VariableDefinition {
	return model.VariableDefinition{VariableName: name, Events: events}
}

func variableReference(name string) model.VariableReference {
	return model.VariableReference{VariableName: name}
}

func variableNameCheck(name string) parseTestCase {
	return parseTestCase{
		label: fmt.Sprintf("variable definition and reference: %s", name),
		given: fmt.Sprintf("%[1]s = r \n%[1]s", name),
		expectUpdates: []model.ScoreUpdate{
			variableDefinition(name, model.Rest{}),
			variableReference(name),
		},
	}
}

func TestVariables(t *testing.T) {
	executeParseTestCases(
		t,
		variableNameCheck("aa"),
		variableNameCheck("aaa"),
		variableNameCheck("HI"),
		variableNameCheck("celloPart2"),
		variableNameCheck("xy42"),
		variableNameCheck("my20cats"),
		variableNameCheck("apple_cider"),
		variableNameCheck("underscores_are_great"),
		parseTestCase{
			label: "variable reference in a part",
			given: "flan = e\nflute: c flan f",
			expectUpdates: []model.ScoreUpdate{
				variableDefinition("flan",
					model.Note{
						Pitch: model.LetterAndAccidentals{NoteLetter: model.E},
					},
				),
				model.PartDeclaration{Names: []string{"flute"}},
				model.Note{Pitch: model.LetterAndAccidentals{NoteLetter: model.C}},
				variableReference("flan"),
				model.Note{Pitch: model.LetterAndAccidentals{NoteLetter: model.F}},
			},
		},
		parseTestCase{
			label: "only a variable reference in a part",
			given: "pudding123 = r\nclarinet: pudding123",
			expectUpdates: []model.ScoreUpdate{
				variableDefinition("pudding123", model.Rest{}),
				model.PartDeclaration{Names: []string{"clarinet"}},
				variableReference("pudding123"),
			},
		},
		parseTestCase{
			label: "variable definition containing a cram expression",
			given: "cheesecake = { c/e }2",
			expectUpdates: []model.ScoreUpdate{
				variableDefinition(
					"cheesecake",
					model.Cram{
						Events: []model.ScoreUpdate{
							model.Chord{
								Events: []model.ScoreUpdate{
									model.Note{
										Pitch: model.LetterAndAccidentals{NoteLetter: model.C},
									},
									model.Note{
										Pitch: model.LetterAndAccidentals{NoteLetter: model.E},
									},
								},
							},
						},
						Duration: model.Duration{
							Components: []model.DurationComponent{
								model.NoteLength{Denominator: 2},
							},
						},
					},
				),
			},
		},
		parseTestCase{
			label: "variable definition within an instrument part",
			given: `harpsichord:
			custard_ = c d e/g`,
			expectUpdates: []model.ScoreUpdate{
				model.PartDeclaration{Names: []string{"harpsichord"}},
				variableDefinition(
					"custard_",
					model.Note{Pitch: model.LetterAndAccidentals{NoteLetter: model.C}},
					model.Note{Pitch: model.LetterAndAccidentals{NoteLetter: model.D}},
					model.Chord{
						Events: []model.ScoreUpdate{
							model.Note{
								Pitch: model.LetterAndAccidentals{NoteLetter: model.E},
							},
							model.Note{
								Pitch: model.LetterAndAccidentals{NoteLetter: model.G},
							},
						},
					},
				),
			},
		},
		parseTestCase{
			label: "variable definition within an instrument part (variation)",
			given: `glockenspiel:
		
			sorbet=c d e/g
			c`,
			expectUpdates: []model.ScoreUpdate{
				model.PartDeclaration{Names: []string{"glockenspiel"}},
				variableDefinition(
					"sorbet",
					model.Note{Pitch: model.LetterAndAccidentals{NoteLetter: model.C}},
					model.Note{Pitch: model.LetterAndAccidentals{NoteLetter: model.D}},
					model.Chord{
						Events: []model.ScoreUpdate{
							model.Note{
								Pitch: model.LetterAndAccidentals{NoteLetter: model.E},
							},
							model.Note{
								Pitch: model.LetterAndAccidentals{NoteLetter: model.G},
							},
						},
					},
				),
				model.Note{Pitch: model.LetterAndAccidentals{NoteLetter: model.C}},
			},
		},
		parseTestCase{
			label: "variable definition before an instrument part",
			given: `GELATO=d e
		
			clavinet: c/f`,
			expectUpdates: []model.ScoreUpdate{
				variableDefinition(
					"GELATO",
					model.Note{Pitch: model.LetterAndAccidentals{NoteLetter: model.D}},
					model.Note{Pitch: model.LetterAndAccidentals{NoteLetter: model.E}},
				),
				model.PartDeclaration{Names: []string{"clavinet"}},
				model.Chord{
					Events: []model.ScoreUpdate{
						model.Note{Pitch: model.LetterAndAccidentals{NoteLetter: model.C}},
						model.Note{Pitch: model.LetterAndAccidentals{NoteLetter: model.F}},
					},
				},
			},
		},
		// Regression test for https://github.com/alda-lang/alda-core/issues/52
		// which involved incorrectly parsing a rest `r` followed by a newline
		// within a variable definition.
		parseTestCase{
			label: "var defined and used s/t var ends with a rest",
			given: "foo = c8 d c r\npiano: foo*2",
			expectUpdates: []model.ScoreUpdate{
				variableDefinition(
					"foo",
					model.Note{
						Pitch: model.LetterAndAccidentals{NoteLetter: model.C},
						Duration: model.Duration{
							Components: []model.DurationComponent{
								model.NoteLength{Denominator: 8},
							},
						},
					},
					model.Note{Pitch: model.LetterAndAccidentals{NoteLetter: model.D}},
					model.Note{Pitch: model.LetterAndAccidentals{NoteLetter: model.C}},
					model.Rest{},
				),
				model.PartDeclaration{Names: []string{"piano"}},
				model.Repeat{Event: variableReference("foo"), Times: 2},
			},
		},
		// Regression test for https://github.com/alda-lang/alda-core/issues/64
		// NB: the trailing newline was essential to reproducing the issue!
		parseTestCase{
			label: "variable definition ending with a variable reference",
			given: `
			soprano = r
			alto = r
			tenor = r
			bass = r
			satb = V1: soprano V2: alto V3: tenor V4: bass
			`,
			expectUpdates: []model.ScoreUpdate{
				variableDefinition("soprano", model.Rest{}),
				variableDefinition("alto", model.Rest{}),
				variableDefinition("tenor", model.Rest{}),
				variableDefinition("bass", model.Rest{}),
				variableDefinition(
					"satb",
					model.VoiceMarker{VoiceNumber: 1},
					variableReference("soprano"),
					model.VoiceMarker{VoiceNumber: 2},
					variableReference("alto"),
					model.VoiceMarker{VoiceNumber: 3},
					variableReference("tenor"),
					model.VoiceMarker{VoiceNumber: 4},
					variableReference("bass"),
				),
			},
		},
		// Regression test for https://github.com/alda-lang/alda-core/issues/64
		// NB: the trailing newline was essential to reproducing the issue!
		parseTestCase{
			label: "variable definition followed by a newline",
			given: `bar = r
			foo = bar
			`,
			expectUpdates: []model.ScoreUpdate{
				variableDefinition("bar", model.Rest{}),
				variableDefinition("foo", variableReference("bar")),
			},
		},
	)
}
